src/java/src/main/java/com/dnanexus/DXSearch.java

```diff
@@ -989,14 +1022,110 @@ private T getDataObjectInstanceFromResult(FindDataObjectsResponse.Entry e) {
             return (T) dataObject;
         }
 
-
         @Override
         public Iterator<T> iterator() {
             return new ResultIterator();
         }
     }
 
     /**
+     * The page subset of data objects that matched a {@code findDataObjects} query.
+     *
+     * @param <T> data object class to be returned
+     */
+    public static class SearchPage<T extends DXDataObject> implements Iterable<T> {
+        private final String classConstraint;
+        private final DXEnvironment env;
+        private final FindDataObjectsResponse response;
+
+        /**
+         * Iterator implementation for findDataObjects results page items.
+         */
+        @VisibleForTesting
+        private class ResultIterator implements Iterator<T> {
+
+            int nextElementIndex;
+
+            private ResultIterator() {
+                this.nextElementIndex = 0;
+            }
+
+            @Override
+            public boolean hasNext() {
+                return nextElementIndex < response.results.size();
+            }
+
+            @Override
+            public T next() {
+                return getDataObjectInstanceFromResult(response.results.get(nextElementIndex++));
+            }
+
+            @Override
+            public void remove() {
+                throw new UnsupportedOperationException();
+            }
+
+            @SuppressWarnings("unchecked")
+            private T getDataObjectInstanceFromResult(FindDataObjectsResponse.Entry e) {
```

Comment:
> Can this be factored out so as to not duplicate the same method in FindDataObjectsResult.ResultIterator?

Since **getDataObjectInstanceFromResult()** is a member of **FindDataObjectsResult<T>** class I decided to move **SearchPage** class there. And I've renamed it to **FindDataObjectsResult<T>::Page** because the name of outer class (**FindDataObjectsResult**) contains all necessary information about what this **Page** is about. Keep me informed, please, if you are not comfortable with such naming.

> Once you've done so, consider whether Guava's Iterators.transform would be useful here (it seems like it might obviate the need for this ResultIterator class you've defined).

I looked to Guava library and found it very interesting and helpful - thank you very much! So I've removed a **ResultIterator** class from the code and implemented that using Guava's **AbstractIterator**.